src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.UnifiedBuild.Tasks/JoinVerticals.cs

```diff
@@ -159,86 +173,87 @@ private async Task ExecuteAsync()
+            await DownloadFileFromArtifact(
+                filesInformation, 
+                artifactName, 
+                azureDevOpsClient, 
+                buildId, 
+                fileName, 
+                outputDirectory, 
+                clientThrottle)));
+    }
+
+    private async Task DownloadFileFromArtifact(
+        ArtifactFiles artifactFilesMetadata, 
+        string azureDevOpsArtifact, 
+        AzureDevOpsClient azureDevOpsClient, 
+        string buildId, 
+        string manifestFile, 
+        string destinationDirectory,
+        SemaphoreSlim clientThrottle)
+    {
         try
         {
             await clientThrottle.WaitAsync();
+            
+            ArtifactItem fileItem;
 
-            ArtifactFiles filesInformation = await azureDevOpsClient.GetArtifactFilesInformation(buildId, artifactName, _retryCount);
+            var matchingFilePaths = artifactFilesMetadata.Items.Where(f => Path.GetFileName(f.Path) == Path.GetFileName(manifestFile));
 
-            foreach (var package in packageFileNames)
+            if (!matchingFilePaths.Any())
             {
-                await DownloadFileFromArtifact(filesInformation, artifactName, azureDevOpsClient, buildId, package, packagesOutputPath);
+                throw new ArgumentException($"File {manifestFile} not found in source files.");
             }
 
-            foreach (var asset in assetFileNames)
+            if (matchingFilePaths.Count() > 1)
             {
-                await DownloadFileFromArtifact(filesInformation, artifactName, azureDevOpsClient, buildId, asset, assetsOutputPath);
-            }
-        }
-        finally
-        {
-            clientThrottle.Release();
-        }
-    }
-
-    private async Task DownloadFileFromArtifact(ArtifactFiles artifactFilesMetadata, string azureDevOpsArtifact, AzureDevOpsClient azureDevOpsClient, string buildId, string manifestFile, string destinationDirectory)
-    {
-        ArtifactItem fileItem;
-
-        var matchingFilePaths = artifactFilesMetadata.Items.Where(f => Path.GetFileName(f.Path) == Path.GetFileName(manifestFile));
-
-        if (!matchingFilePaths.Any())
-        {
-            throw new ArgumentException($"File {manifestFile} not found in source files.");
-        }
-
-        if (matchingFilePaths.Count() > 1)
-        {
-            // Picking the first one until https://github.com/dotnet/source-build/issues/4596 is resolved
-            if (manifestFile.Contains("productVersion.txt"))
-            {
-                fileItem = matchingFilePaths.First();
+                // Picking the first one until https://github.com/dotnet/source-build/issues/4596 is resolved
+                if (manifestFile.Contains("productVersion.txt"))
+                {
+                    fileItem = matchingFilePaths.First();
+                }
+                else
+                {
+                    fileItem = matchingFilePaths
```

Comment:
I'm not sure I understand the purpose of this. It seems to be roughly redundant with the code just above https://github.com/dotnet/sdk/pull/43627/files/82c18d04336d56e52f0df8fe7f18e40999216437..717bfcb8325cd132be588c243981e9d30031f53f#diff-0bb02635d68274012f0c2cabbf5fa8e513c3472b0d93e8bd5b1f14a379741fb5R215

Reply:
If it's not redundant, can you add a comment as to its purpose?

Reply:
I added a comment, this is for the sdk zip which exists in 2 places in the artifact - for example `Windows_x64_Artifacts/assets/Release/dotnet-sdk-*-win-x64.zip` and `Windows_x64_Artifacts/assets/Release/Sdk/*/dotnet-sdk-*-win-x64.zip` but it's listed only once in the vertical manifest
I assume it's getting copied somewhere after the manifest creation but couldn't find where
Comparing the full path solves the issue

Reply:
Gotcha, and my guess is that it's not enough to initially compare with the full path in all cases?

We need to add some tests for this merging code to ensure that the desired selection is being made: https://github.com/dotnet/source-build/issues/4661

Reply:
I found some places where `Contains` doesn't work - both `Microsoft.DotNet.Wpf.GitHub.9.0.0-rc.1.24413.1.nupkg` and `runtime.win-x86.Microsoft.DotNet.Wpf.GitHub.9.0.0-rc.1.24413.1.nupkg` contain the same package name
So we'd need to check both that the filename fully matches and compare the path, since we need to check the path only in few cases it made sense to do it this way